Tests/UpstreamIntegrationTests/util/XCTestCase+AnyCodableAsserts.swift

```diff
@@ -0,0 +1,729 @@
+    ///
+    /// There are 3 different ways to specify alternate mode paths for arrays:
+    /// 1. The specific index: [\<INT\>] (ex: `[0]`, `[28]`, etc.) - The element at the specified index will use the alternate mode.
+    /// 2. The wildcard index: [*\<INT\>] (ex: `[*1]`, `[*12]`, etc) - The element at the specified index will use the alternate mode and apply wildcard matching logic.
+    /// 3. The general wildcard: [*] (must be in exactly this format) - Every element not explicitly specified by 1 or 2 will use the alternate mode and apply wildcard matching logic. This option is mututally exclusive with default behavior.
+    /// - The default behavior is that elements from the expected JSON side are compared in order, up to the last element of the expected array.
+    ///
+    /// - Parameters:
+    ///    - expected: The expected JSON in AnyCodable format used to perform the assertions
+    ///    - actual: The actual JSON in AnyCodable format that is validated against `expected`
+    ///    - typeMatchPaths: Optionally, the key paths in the expected JSON that should use type matching mode, where values require only the same type (and non-nil, given the expected value is not `nil` itself)
+    ///    - file: the file to show test assertion failures in
+    ///    - line: the line to show test assertion failures on
+    func assertExactMatch(expected: AnyCodable?, actual: AnyCodable?, typeMatchPaths: [String] = [], file: StaticString = #file, line: UInt = #line) {
+        let pathTree = generatePathTree(paths: typeMatchPaths)
+        assertFlexibleEqual(expected: expected, actual: actual, pathTree: pathTree, exactMatchMode: true, file: file, line: line)
+    }
+
+    /// Performs flexible comparison testing assertions between two `AnyCodable` instances.
+    @discardableResult
+    private func assertFlexibleEqual(expected: AnyCodable?, actual: AnyCodable?, keyPath: [Any] = [], pathTree: [String: Any]?, exactMatchMode: Bool, file: StaticString = #file, line: UInt = #line, shouldAssert: Bool = true) -> Bool {
+        if expected?.value == nil {
+            return true
+        }
+        guard let expected = expected, let actual = actual else {
+            if shouldAssert {
+                XCTFail(#"""
+                    Expected JSON is non-nil but Actual JSON is nil.
+
+                    Expected: \#(String(describing: expected))
+
+                    Actual: \#(String(describing: actual))
+
+                    Key path: \#(keyPathAsString(keyPath))
+                """#, file: file, line: line)
+            }
+            return false
+        }
+
+        switch (expected, actual) {
+        case let (expected, actual) where (expected.value is String && actual.value is String):
+            fallthrough
+        case let (expected, actual) where (expected.value is Bool && actual.value is Bool):
+            fallthrough
```

Comment:
why do we need to still `fallthrough` if an exact type match was already identified?